File: src/core/engine/expression-evaluator.js
const OPERATORS = {
  "+": { precedence: 1, apply: (left, right) => left + right },
  "-": { precedence: 1, apply: (left, right) => left - right },
  "*": { precedence: 2, apply: (left, right) => left * right },
  "/": { precedence: 2, apply: (left, right) => left / right },
};

export function createFormulaContext(characterProjection) {
  const abilityModifiers = characterProjection?.abilityModifiers ?? {};
  return {
    prof: characterProjection?.proficiencyBonus ?? 0,
    proficiency: characterProjection?.proficiencyBonus ?? 0,
    str_mod: abilityModifiers.str ?? 0,
    dex_mod: abilityModifiers.dex ?? 0,
    con_mod: abilityModifiers.con ?? 0,
    int_mod: abilityModifiers.int ?? 0,
    wis_mod: abilityModifiers.wis ?? 0,
    cha_mod: abilityModifiers.cha ?? 0,
  };
}

export function evaluateFormula(formula, context = {}) {
  if (typeof formula === "number") return formula;
  if (!formula || typeof formula !== "string") return 0;

  const tokens = tokenize(formula);
  const output = [];
  const operators = [];

  tokens.forEach((token) => {
    if (token.type === "number" || token.type === "variable") {
      output.push(token);
      return;
    }

    if (token.value === "(") {
      operators.push(token);
      return;
    }

    if (token.value === ")") {
      while (operators.length && operators.at(-1).value !== "(") output.push(operators.pop());
      if (!operators.length) throw new Error(`Invalid formula: ${formula}`);
      operators.pop();
      return;
    }

    const current = OPERATORS[token.value];
    while (operators.length) {
      const previous = OPERATORS[operators.at(-1).value];
      if (!previous || previous.precedence < current.precedence) break;
      output.push(operators.pop());
    }
    operators.push(token);
  });

  while (operators.length) {
    const operator = operators.pop();
    if (operator.value === "(") throw new Error(`Invalid formula: ${formula}`);
    output.push(operator);
  }

  return evaluateRpn(output, context);
}

function tokenize(formula) {
  const tokens = [];
  const pattern = /\s*(@[a-zA-Z_][a-zA-Z0-9_]*|\d+(?:\.\d+)?|[()+\-*/])\s*/g;
  let match;
  let cursor = 0;

  while ((match = pattern.exec(formula))) {
    if (match.index !== cursor) throw new Error(`Unsupported formula token near: ${formula.slice(cursor)}`);
    cursor = pattern.lastIndex;
    const value = match[1];
    if (/^\d/.test(value)) tokens.push({ type: "number", value: Number(value) });
    else if (value.startsWith("@")) tokens.push({ type: "variable", value: value.slice(1) });
    else tokens.push({ type: "operator", value });
  }

  if (cursor !== formula.length) throw new Error(`Unsupported formula token near: ${formula.slice(cursor)}`);
  return tokens;
}

function evaluateRpn(tokens, context) {
  const stack = [];

  tokens.forEach((token) => {
    if (token.type === "number") {
      stack.push(token.value);
      return;
    }
    if (token.type === "variable") {
      const value = Number(context[token.value]);
      stack.push(Number.isFinite(value) ? value : 0);
      return;
    }

    const right = stack.pop();
    const left = stack.pop();
    if (!Number.isFinite(left) || !Number.isFinite(right)) throw new Error("Invalid formula expression");
    stack.push(OPERATORS[token.value].apply(left, right));
  });

  if (stack.length !== 1) throw new Error("Invalid formula expression");
  return stack[0];
}

File: backend/app/domain/engine/expression_evaluator.py
"""Minimal formula evaluation helpers ported from the frontend engine."""

import math
import re

OPERATORS = {
    "+": (1, lambda left, right: left + right),
    "-": (1, lambda left, right: left - right),
    "*": (2, lambda left, right: left * right),
    "/": (2, lambda left, right: left / right),
}

TOKEN_PATTERN = re.compile(r"@[a-zA-Z_][a-zA-Z0-9_]*|\d+(?:\.\d+)?|[()+\-*/]")


def evaluate_formula(
    formula: str | int | float, context: dict[str, float] | None = None
) -> float:
    if isinstance(formula, (int, float)):
        return float(formula)
    if not formula:
        return 0

    context = context or {}
    tokens = _tokenize(formula)
    output: list[str] = []
    operators: list[str] = []

    for token in tokens:
        if token.startswith("@") or token.replace(".", "", 1).isdigit():
            output.append(token)
            continue
        if token == "(":
            operators.append(token)
            continue
        if token == ")":
            while operators and operators[-1] != "(":
                output.append(operators.pop())
            if not operators:
                raise ValueError(f"Invalid formula: {formula}")
            operators.pop()
            continue

        precedence = OPERATORS[token][0]
        while (
            operators
            and operators[-1] in OPERATORS
            and OPERATORS[operators[-1]][0] >= precedence
        ):
            output.append(operators.pop())
        operators.append(token)

    while operators:
        operator = operators.pop()
        if operator == "(":
            raise ValueError(f"Invalid formula: {formula}")
        output.append(operator)

    stack: list[float] = []
    for token in output:
        if token.startswith("@"):
            stack.append(_numeric_value(context.get(token[1:], 0)))
        elif token.replace(".", "", 1).isdigit():
            stack.append(float(token))
        else:
            if len(stack) < 2:
                raise ValueError("Invalid formula expression")
            right = stack.pop()
            left = stack.pop()
            stack.append(OPERATORS[token][1](left, right))

    if len(stack) != 1:
        raise ValueError("Invalid formula expression")

    return stack[0]


def _tokenize(formula: str) -> list[str]:
    tokens: list[str] = []
    cursor = 0

    for match in TOKEN_PATTERN.finditer(formula):
        if formula[cursor:match.start()].strip():
            fragment = formula[cursor:]
            raise ValueError(f"Unsupported formula token near: {fragment}")
        tokens.append(match.group())
        cursor = match.end()

    if formula[cursor:].strip():
        fragment = formula[cursor:]
        raise ValueError(f"Unsupported formula token near: {fragment}")

    return tokens


def _numeric_value(value: object) -> float:
    try:
        number = float(value)
    except (TypeError, ValueError):
        return 0.0
    return number if math.isfinite(number) else 0.0
